libs/graphics/lcd_spi_unbuf.c

```diff
@@ -0,0 +1,189 @@
+
+#include "lcd_spi_unbuf.h"
+#include "jsutils.h"
+#include "jsinteractive.h"
+#include "jswrap_graphics.h"
+#include "jshardware.h"
+
+int _pin_mosi;
+int _pin_clk;
+int _pin_cs;
+int _pin_dc;
+int _colstart;
+int _rowstart;
+IOEventFlags _device;
+
+static void spi_cmd(const uint8_t cmd) 
+{
+  jshPinSetValue(_pin_dc, 0);
+  jshSPISend(_device, cmd);
+  jshPinSetValue(_pin_dc, 1);
+}
+
+static void spi_data(const uint8_t *data, int len) 
+{
+  if (len==0) return;
+  jshSPISendMany(_device, data, NULL, len, NULL);
+}
+
+void jshLCD_SPI_UNBUFInitInfo(JshLCD_SPI_UNBUFInfo *inf) {
+  inf->pinCS         = PIN_UNDEFINED;
+  inf->pinDC         = PIN_UNDEFINED;
+  inf->width         = 240;
+  inf->height        = 320;
+  inf->colstart        = 0;
+  inf->rowstart        = 0;
+}
+
+bool jsspiPopulateOptionsInfo( JshLCD_SPI_UNBUFInfo *inf, JsVar *options){
+  jshLCD_SPI_UNBUFInitInfo(inf);
+  jsvConfigObject configs[] = {
+    {"cs", JSV_PIN, &inf->pinCS},
+    {"dc", JSV_PIN, &inf->pinDC},
+    {"width", JSV_INTEGER , &inf->width},
+    {"height", JSV_INTEGER , &inf->height},
+    {"colstart", JSV_INTEGER , &inf->colstart},
+    {"rowstart", JSV_INTEGER , &inf->rowstart},
+  };  
+  
+  if (jsvReadConfigObject(options, configs, sizeof(configs) / sizeof(jsvConfigObject))) {	
+    if ( inf->pinDC == PIN_UNDEFINED  || inf->pinCS == PIN_UNDEFINED ) {
+      return false;
+    }
+    return true;
+  }  
+  else { 
+    return false;
+  }
+}
+
+/*JSON{
+  "type" : "staticmethod",
+  "class" : "lcd_spi_unbuf",
+  "name" : "connect",
+  "generate" : "jswrap_lcd_spi_unbuf_connect",
+  "params" : [
+    ["device","JsVar","The used SPI device"],
+    ["options","JsVar","An Object containing extra information"]
+  ],
+  "return" : ["JsVar","The new Graphics object"],
+  "return_object" : "Graphics"
+}*/
+JsVar *jswrap_lcd_spi_unbuf_connect(JsVar *device, JsVar *options) { 
+  JsVar *parent = jspNewObject(0, "Graphics");
+  if (!parent) {
+    jsExceptionHere(JSET_ERROR,"creating new object Graphics");
+    return NULL;  
+  }
+
+  JshLCD_SPI_UNBUFInfo inf;
+
+  if (!jsspiPopulateOptionsInfo(&inf, options)) {
+    jsExceptionHere(JSET_ERROR,"pins not supplied correctly");
+    jsvUnLock(parent);
+  	return NULL;
+  }
+  _pin_cs = inf.pinCS;
+  _pin_dc = inf.pinDC;
+  _colstart = inf.colstart;
+  _rowstart = inf.rowstart;
+  _device = jsiGetDeviceFromClass(device);
+	
+  if (!DEVICE_IS_SPI(_device)) { 
+  	jsExceptionHere(JSET_ERROR,"Software SPI is not supported for now");
+  	jsvUnLock(parent);
+    return NULL;
+  }  
+
+  JsGraphics gfx;
+  graphicsStructInit(&gfx,inf.width,inf.height,16);
+  gfx.data.type = JSGRAPHICSTYPE_LCD_SPI_UNBUF;
+  gfx.graphicsVar = parent;
+  
+  jshPinOutput(_pin_dc, 1);
+  jshPinSetValue(_pin_dc, 1);
+
+  jshPinOutput(_pin_cs, 1);
+  jshPinSetValue(_pin_cs, 1);
+  
+  lcd_spi_unbuf_setCallbacks(&gfx);
+  graphicsSetVar(&gfx);	
+  return parent;
+}
+
+void disp_spi_transfer_addrwin(int x1, int y1, int x2, int y2) {
+  uint32_t wd;
```

Comment:
This looks good, but wouldn't it be easier just to use a `unsigned char wd[4]` here? The whole shifting into a 32 bit number just seems a bit involved and hard to read?

Reply:
It would also work yes, 

it should also be possible to use an uint16_t wd[2] so no shifting is needed at all.

that way its even more readable. 

not tested right now.

sometimes you dont see it because you looked so long onto it^^